### PR TITLE
fix(security): gate PersonaService preference save & document authz-primitive endpoints (#2798)

### PR DESCRIPTION
**Pull request by Ben-Heerema**
Part of #2798 (the umbrella issue spans ~8 JAX-RS services and is intentionally kept open — this PR uses **Part of**, not `fixes`, so the umbrella is not auto-closed).

## Why PersonaService is special
PersonaService is the carved-out "authorization primitive" case of #2798. Several of its endpoints **are** the privilege API the UI calls to decide what to render — so the blanket "add `hasPrivilege()` everywhere" remediation is the wrong mental model here. This PR handles it per-endpoint instead of sweeping.

## Scope
- **Gate the one genuine gap:** `saveMyPatientListConfig` (`POST /persona/patientList/config`) persists the caller's own provider preferences (`patientListConfig.*`) via the same `UserProperty` mechanism as its sibling mutators `updatePreference`/`updatePreferences`, but had no privilege check. Add the same `_pref`/`u` check for parity.
- **Document, don't gate, the self-scoped authorization primitives** (`/rights`, `/hasRight`, `/hasRights`, `/isAllowedAccessToPatientRecord`): each resolves its subject from `getLoggedInInfo()` and only ever reports the *current* caller's own rights. Gating them would be circular and break the UI's permission rendering, with no IDOR to defend. They already fail closed when unauthenticated (`getLoggedInInfo()` throws with no `LoggedInInfo`).
- **Document why `setDefaultProgram` needs no secobj:** it is self-scoped (own `providerNo`) and `ProgramManager2.setCurrentProgramInDomain` already rejects a `programId` outside the provider's domain; the program admin secobjs (`_pmm_management`/`_program*`/`_programEdit`) govern administering programs, not a provider switching their own active program.
- The other own-data read endpoints (`/navbar`, `/patientLists`, `/patientList/config` GET, `/preferences`, `/dashboardMenu`) are per-provider self-scoped reads and remain open by design.

## Privilege-object choice
`_pref`/`u` matches the two sibling preference mutators in the same class. (The broader codebase gates provider-preference *writes* with `_pref`/`w`; in-file consistency was preferred — happy to switch to `w` if reviewers prefer the stricter codebase-wide convention.)

## Tests
`PersonaServiceUnitTest` (new) — `@Tag("unit")`:
- allow path: with `_pref`/`u` granted, both provider properties are persisted and the privilege check is invoked.
- deny path: with `_pref`/`u` denied, the method throws `Access Denied` and persists nothing.

## Local verification
- `mvn -o -Dtest=PersonaServiceUnitTest test` → Tests run: 2, Failures: 0, Errors: 0.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Gate PersonaService provider preference saves with a privilege check while documenting why other self-scoped endpoints remain ungated.

Bug Fixes:
- Require _pref/u privilege before saving patient list configuration for the logged-in provider to close an authorization gap.

Enhancements:
- Clarify in-code rationale for leaving self-scoped authorization-primitive endpoints and default-program switching ungated by security objects.

Tests:
- Add PersonaService unit tests that cover allowed and denied paths for saving patient list configuration, including boundary behavior for appointment count persistence.

---
## Summary by cubic
Secures provider preference saves in PersonaService by requiring `_pref`/`u`, returning a 403 via `SecurityException` when denied. Documents why self-scoped authz endpoints remain open to avoid breaking UI permission checks. Part of #2798.

- **Bug Fixes**
  - Require `_pref`/`u` for `saveMyPatientListConfig` (`POST /persona/patientList/config`), matching `updatePreference`/`updatePreferences`.
  - On denied privilege, throw `SecurityException("missing required sec object (_pref)")`; tests capture the caller’s `LoggedInInfo`, assert exact saved properties, and in the boundary case assert only `patientListConfig.showReason=true` is persisted when the count is not > 0.
  - Document that `/rights`, `/hasRight`, `/hasRights`, and `/isAllowedAccessToPatientRecord` are self-scoped and intentionally ungated; `setDefaultProgram` is also self-scoped and validated by `ProgramManager2.setCurrentProgramInDomain`.

<sup>Written for commit 3a687f5e1a811a07e527e197e3cae68ebc202b1f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/carlos-emr/carlos/pull/3091?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

File: src/main/java/io/github/carlos_emr/carlos/webserv/rest/PersonaService.java [hunk 1]
```diff
@@ -356,6 +372,13 @@ public PatientListConfigTo1 getMyPatientListConfig() {
     public PatientListConfigTo1 saveMyPatientListConfig(PatientListConfigTo1 patientListConfigTo1) {
         Provider provider = getCurrentProvider();
 
+        // Persists the caller's own provider preferences (patientListConfig.*), the same
+        // UserProperty mechanism guarded by updatePreference/updatePreferences below — gate it
+        // with the same _pref/u privilege for parity (#2798).
+        if (!securityInfoManager.hasPrivilege(getLoggedInInfo(), "_pref", "u", null)) {
+            throw new RuntimeException("Access Denied");
+        }
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/webserv/rest/PersonaService.java:380]: Addressed on current head (21240920): PersonaService.java:379 now throws SecurityException("missing required sec object (_pref)") — the CARLOS paren-form convention, matching MeasurementService/LabService in the same REST stack. Resolving.

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        when(mockSecurityInfoManager.hasPrivilege(any(), eq("_pref"), eq("u"), isNull())).thenReturn(true);
+        when(mockUserPropertyDao.getProp(eq(PROVIDER_NO), any())).thenReturn(null);
+
+        PatientListConfigTo1 config = new PatientListConfigTo1();
+        config.setNumberOfApptstoShow(10);
+        config.setShowReason(true);
+
+        service.saveMyPatientListConfig(config);
+
+        verify(mockSecurityInfoManager).hasPrivilege(any(), eq("_pref"), eq("u"), isNull());
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java:161]: Done in 9e53137f: the allow-path test now captures the LoggedInInfo and asserts isSameAs(loggedInInfo), validating the privilege check uses the current callers own context.

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        // numberOfApptsToShow (>0) and showReason are each saved.
+        verify(mockUserPropertyDao, org.mockito.Mockito.times(2)).saveProp(any(io.github.carlos_emr.carlos.commn.model.UserProperty.class));
+    }
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java:167]: Partially adopted in 9e53137f: added shouldSkipAppointmentCount_whenNotPositive (numberOfApptstoShow not >0 means the count branch is skipped, _pref/u still enforced). Did NOT add the showReason=null cases (isShowReason() is a primitive boolean, so setShowReason(null) would not compile) nor the all-fields-unset to never(saveProp) case (showReason is always persisted, so that expectation is incorrect). Resolving.

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+        service.saveMyPatientListConfig(config);
+
+        // The privilege check must be made against the current caller's own context, not some
+        // other subject — capture and assert the LoggedInInfo rather than matching any().
+        ArgumentCaptor<LoggedInInfo> infoCaptor = ArgumentCaptor.forClass(LoggedInInfo.class);
+        verify(mockSecurityInfoManager).hasPrivilege(infoCaptor.capture(), eq("_pref"), eq("u"), isNull());
+        assertThat(infoCaptor.getValue()).isSameAs(loggedInInfo);
+        // numberOfApptsToShow (>0) and showReason are each saved.
+        verify(mockUserPropertyDao, times(2)).saveProp(any(UserProperty.class));
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java]: Done in c6733782: the allow-path test now captures the persisted UserProperty args and asserts provider/name/value via containsExactlyInAnyOrder (patientListConfig.numberOfApptsToShow=10 and patientListConfig.showReason=true for provider 101).

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

File: src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+        PatientListConfigTo1 config = new PatientListConfigTo1();
+        config.setNumberOfApptstoShow(0); // not > 0 → count is not persisted
+        config.setShowReason(true);
+
+        service.saveMyPatientListConfig(config);
+
+        // privilege is still enforced, and only showReason is persisted (count branch skipped).
+        verify(mockSecurityInfoManager).hasPrivilege(any(), eq("_pref"), eq("u"), isNull());
+        verify(mockUserPropertyDao, times(1)).saveProp(any(UserProperty.class));
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/webserv/rest/PersonaServiceUnitTest.java]: Done in 3a687f5e: the boundary test now captures the single persisted UserProperty and asserts containsExactly(PROVIDER_NO, patientListConfig.showReason, true), so it fails if the count were saved instead of skipped.

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@yingbull ready for review

**Review by yingbull — APPROVED** [review]
